libs/langchain/langchain/embeddings/vertexai.py

```diff
@@ -1,46 +1,271 @@
+            "CLASSIFICATION",
+            "CLUSTERING",
+        ]
+    ] = None
+
+    # instance context
+    instance: Dict[str, Any] = {}  #: :meta private:
+
     @root_validator()
     def validate_environment(cls, values: Dict) -> Dict:
         """Validates that the python package exists in environment."""
         cls._try_init_vertexai(values)
         try:
             from vertexai.language_models import TextEmbeddingModel
+
+            values["client"] = TextEmbeddingModel.from_pretrained(values["model_name"])
         except ImportError:
             raise_vertex_import_error()
-        values["client"] = TextEmbeddingModel.from_pretrained(values["model_name"])
         return values
 
+    def __init__(
+        self,
+        project: Optional[str] = None,
+        location: str = "us-central1",
+        request_parallelism: int = 5,
+        max_retries: int = 6,
+        model_name: str = "textembedding-gecko",
+        credentials: Optional[Any] = None,
+        embeddings_type: Optional[
+            Literal[
+                "RETRIEVAL_QUERY",
+                "RETRIEVAL_DOCUMENT",
+                "SEMANTIC_SIMILARITY",
+                "CLASSIFICATION",
+                "CLUSTERING",
+            ]
+        ] = None,
+        **kwargs: Any,
+    ):
+        """Initialize the sentence_transformer."""
+        super().__init__(
+            project=project,
+            location=location,
+            credentials=credentials,
+            request_parallelism=request_parallelism,
+            max_retries=max_retries,
+            **kwargs,
+        )
+        self.model_name = model_name
+        self.embeddings_type = embeddings_type
+
+        self.instance["batch_size"] = _MAX_BATCH_SIZE
+        self.instance["min_good_batch_size"] = _MIN_BATCH_SIZE
+        self.instance["lock"] = threading.Lock()
+        self.instance["batch_size_validated"] = False
+        self.instance["task_executor"] = ThreadPoolExecutor(
+            max_workers=request_parallelism
+        )
+
+    @staticmethod
+    def _split_by_punctuation(text: str) -> List[str]:
+        """Splits a string by punctuation and whitespace characters."""
+        split_by = string.punctuation + "\t\n "
+        pattern = f"([{split_by}])"
+        # Using re.split to split the text based on the pattern
+        return [segment for segment in re.split(pattern, text) if segment]
+
+    def _prepare_batches(
+        self, texts: List[str], batch_size: int = 0
+    ) -> List[List[str]]:
+        """Splits texts in batches based on current maximum batch size
+        and maximum tokens per request.
+        """
+        if batch_size == 0:
+            batch_size = self.instance["batch_size"]
+        text_index = 0
+        texts_len = len(texts)
+        batch_token_len = 0
+        batches: List[List[str]] = []
+        current_batch: List[str] = []
+        if texts_len == 0:
+            return []
+        while text_index < texts_len:
```

Comment:
it looks to me, that if the length of one element is above _MAX_TOKENS_PER_BATCH, the script will run forever.

and since it's conservative estimation, we always should have at least one element per batch (even if this element exceed the conservative limit) and let the model fail during inference with an appropriate error.

Reply:
Great catch. Thank you! Done.